// coverflow/coverflow.js
var $ = global.$
var _ = global._
var Backbone = global.Backbone
var Fs = require('fs')

var CoverflowModel = Backbone.Model.extend({
  defaults:{
    width:480,
    height:270,
    cellWidth:180,
    cellHeight:180,
    selected:0,
    perspective:250,
    coverGap:40,
    coverOffset:130,
    collection:new Backbone.Collection    
  }
})

var CoverflowView = Backbone.View.extend({

  initialize : function(){
    if(!this.template){
      this.template = _.template(Fs.readFileSync(__dirname+'/default-template.html').toString())
    }
        
    this.initFromModel()
    this.createDOM()
    this.initBindings()
  },
  
  initFromModel : function(){
    this.collection = this.model.get('collection')    
    this.selected = this.collection.at(this.model.get('selected'))
    this.cellWidth = this.model.get('cellWidth')
    this.cellHeight = this.model.get('cellHeight')
    this.height = this.model.get('height')
    this.width = this.model.get('width')
    this.perspective = this.model.get('perspective')
    this.coverGap = this.model.get('coverGap')
    this.coverOffset = this.model.get('coverOffset')
  },
  
  initBindings : function(){
    _.bindAll(this, 'render')    
    this.on('selectedChange', this.render)    
  },
  
  select : function(iIndex){
    if(iIndex >= 0 && iIndex < this.collection.length){
      var newSelected = this.collection.at(iIndex)
      if(newSelected != this.selected){      
        this.selected = newSelected
        this.trigger('selectedChange')
      }
    }
  },
  
  Next : function(){
    var next = null
    if(this.selected == null){
      this.select(0)
    }else{
      var indexSelected = this.collection.indexOf(this.selected)
      this.select( this.getNextIndex(indexSelected) )
    }
  },

  Previous : function(){
    var previous = null
    if(this.selected == null){
      this.select(0)
    }else{
      var indexSelected = this.collection.indexOf(this.selected)
      this.select( this.getPreviousIndex(indexSelected) )
    }    
  },
  
  getNextIndex : function(baseIndex){
    return (baseIndex + 1) % this.collection.size()
  },
  
  getPreviousIndex : function(baseIndex){
    return (baseIndex + this.collection.size() - 1) % this.collection.size() 
  },  
  
  createDOM : function(){
    this.$divWrap = $(window.document.createElement('div'))
    this.$divTray = $(window.document.createElement('div'))
    this.$divWrap.addClass('coverflow-wrap')
    this.$divTray.addClass('coverflow-tray')
    
    var self = this    
    this.collection.forEach(function(item){
      var $div = $(window.document.createElement('div'))
      $div.html(self.template({item:item}))
      $div.addClass('coverflow-cell')            
      self.$divTray.append($div)
    })
    this.$divWrap.append(this.$divTray)
    this.$divWrap.css('-webkit-perspective', this.perspective+'px')
    this.$divWrap.css('left', (this.width/2)+'px')
    this.$divWrap.css('top', (this.height/2)+'px')
    this.$el.append(this.$divWrap)
    this.$el.addClass('coverflow')
    this.$el.css('width', this.width)
    this.$el.css('height', this.height)
    
  },
  
  render : function(){
    var self = this
    var indexSelected = this.collection.indexOf(this.selected)
    var collectionLength = this.collection.length    
        
    this.$divTray.css('-webkit-transform', 'translate3d(-'+(this.coverGap*indexSelected)+'px , 0px, 0px)')
    for(var i = 0; i< collectionLength; i++){
      var element = $(this.$divTray.children()[i])
      element.css('height',this.cellHeight+'px')
      element.css('width',this.cellWidth+'px')
      element.css('left',-this.cellWidth/2+'px')
      element.css('top',-this.cellHeight/2+'px')
      if(i< indexSelected){
        element.css('-webkit-transform',' translate3d('+(-this.coverOffset+(i*this.coverGap))+'px, 0px, -170px) rotateY(70deg)')
      }else if(i == indexSelected){        
        element.css('-webkit-transform','translate3d('+this.coverGap*i+'px,0px,0px)')
      }else if(i > indexSelected){        
        element.css('-webkit-transform','translate3d('+(this.coverOffset+(i*this.coverGap))+'px, 0px, -170px) rotateY(-70deg)')
      }      
    }
  }
})

exports.CoverflowModel = CoverflowModel
exports.CoverflowView = CoverflowView
